events/dispatcher_entity_change.go

```diff
@@ -0,0 +1,367 @@
+
+	var changeType event.EntityChangeEventType
+	if state.GetChangeType() == boltz.EntityCreated {
+		changeType = event.EntityChangeTypeEntityCreated
+	} else if state.GetChangeType() == boltz.EntityUpdated {
+		changeType = event.EntityChangeTypeEntityUpdated
+	} else if state.GetChangeType() == boltz.EntityDeleted {
+		changeType = event.EntityChangeTypeEntityDeleted
+	}
+
+	isParentEvent := state.IsParentEvent()
+	evt := &event.EntityChangeEvent{
+		Namespace:          event.EntityChangeEventsNs,
+		EventId:            state.GetEventId(),
+		EventType:          changeType,
+		EntityType:         state.GetStore().GetEntityType(),
+		IsParentEvent:      &isParentEvent,
+		Timestamp:          time.Now(),
+		Metadata:           map[string]any{},
+		InitialState:       state.GetInitialState(),
+		FinalState:         state.GetFinalState(),
+		PropagateIndicator: self.network.Dispatcher.IsLeaderOrLeaderless(),
+	}
+
+	changeCtx := change.FromContext(state.GetCtx().Context())
+	if changeCtx == nil {
+		changeCtx = change.New()
+		state.GetCtx().UpdateContext(func(ctx context.Context) context.Context {
+			return changeCtx.AddToContext(ctx)
+		})
+	}
+
+	changeCtx.PopulateMetadata(evt.Metadata)
+
+	for k, v := range self.globalMetadata {
+		evt.Metadata[k] = v
+	}
+
+	self.dispatcher.AcceptEntityChangeEvent(evt)
+	return self.logTxEvent(state)
+}
+
+func (self *entityChangeEventDispatcher) emitRecoveryEvent(eventId string, entityType string) {
+	evt := &event.EntityChangeEvent{
+		Namespace:       event.EntityChangeEventsNs,
+		EventId:         eventId,
+		EntityType:      entityType,
+		EventType:       event.EntityChangeTypeCommitted,
+		Timestamp:       time.Now(),
+		IsRecoveryEvent: true,
+	}
+	self.dispatcher.AcceptEntityChangeEvent(evt)
+}
+
+func (self *entityChangeEventDispatcher) ProcessPostCommit(state boltz.UntypedEntityChangeState) {
+	isParentEvent := state.IsParentEvent()
+	evt := &event.EntityChangeEvent{
+		Namespace:          event.EntityChangeEventsNs,
+		EventId:            state.GetEventId(),
+		EventType:          event.EntityChangeTypeCommitted,
+		EntityType:         state.GetStore().GetEntityType(),
+		Timestamp:          time.Now(),
+		IsParentEvent:      &isParentEvent,
+		PropagateIndicator: self.network.Dispatcher.IsLeaderOrLeaderless(),
+	}
+	self.dispatcher.AcceptEntityChangeEvent(evt)
+	self.notifyFlush()
+}
+
+func (self *entityChangeEventDispatcher) notifyFlush() {
+	select {
+	case self.notifyCh <- struct{}{}:
+	default:
+	}
+}
+
+func (self *entityChangeEventDispatcher) flushLoop() {
+	for {
+		// wait to be notified of an event
+		<-self.notifyCh
+
+		// wait until we've not gotten an event for 5 seconds before cleaning up
+		flushed := false
+		for !flushed {
+			select {
+			case <-self.notifyCh:
+			case <-time.After(5 * time.Second):
+				pfxlog.Logger().Debug("cleaning up entity change events")
+				self.flushCommittedTxEvents(false)
+				flushed = true
+			}
```

Comment:
In case there's a flurry of changes. Each tx will flush the commit event of the previous tx. This way only the last event will need external flushing.